lib/utils/packet.c

```diff
@@ -0,0 +1,398 @@
+	mbuf->l3_len = sizeof(struct rte_ipv6_hdr);
+	mbuf->l4_len =
+		(proto == IPPROTO_UDP ? sizeof(struct rte_udp_hdr)
+				      : sizeof(struct rte_tcp_hdr));
+	mbuf->packet_type =
+		RTE_PTYPE_L2_ETHER | RTE_PTYPE_L3_IPV6 |
+		(proto == IPPROTO_UDP ? RTE_PTYPE_L4_UDP : RTE_PTYPE_L4_TCP);
+
+	struct rte_ether_hdr *eth =
+		rte_pktmbuf_mtod(mbuf, struct rte_ether_hdr *);
+	eth->ether_type = rte_cpu_to_be_16(RTE_ETHER_TYPE_IPV6);
+
+	struct rte_ipv6_hdr *ip = (struct rte_ipv6_hdr *)(eth + 1);
+	ip->proto = proto;
+	ip->payload_len = rte_cpu_to_be_16(
+		proto == IPPROTO_UDP ? sizeof(struct rte_udp_hdr)
+				     : sizeof(struct rte_tcp_hdr)
+	);
+	memcpy(ip->src_addr, src_ip, 16);
+	memcpy(ip->dst_addr, dst_ip, 16);
+
+	if (proto == IPPROTO_UDP) {
+		struct rte_udp_hdr *udp = (struct rte_udp_hdr *)(ip + 1);
+		udp->src_port = rte_cpu_to_be_16(src_port);
+		udp->dst_port = rte_cpu_to_be_16(dst_port);
+		udp->dgram_len = rte_cpu_to_be_16(sizeof(*udp));
+		udp->dgram_cksum = 0;
+	} else { // tcp
+		struct rte_tcp_hdr *tcp = (struct rte_tcp_hdr *)(ip + 1);
+		tcp->src_port = rte_cpu_to_be_16(src_port);
+		tcp->dst_port = rte_cpu_to_be_16(dst_port);
+		tcp->tcp_flags = flags;
+	}
+
+	return mbuf;
+}
+
+int
+fill_packet_net4(
+	struct packet *packet,
+	const uint8_t src_ip[NET4_LEN],
+	const uint8_t dst_ip[NET4_LEN],
+	uint16_t src_port,
+	uint16_t dst_port,
+	uint8_t proto,
+	uint16_t flags
+) {
+	packet->mbuf =
+		make_mbuf4(src_ip, dst_ip, src_port, dst_port, proto, flags);
+	assert(packet->mbuf != NULL);
+	return parse_packet(packet);
+}
+
+// IPv6 in host byte order
+int
+fill_packet_net6(
+	struct packet *packet,
+	const uint8_t src_ip[NET6_LEN],
+	const uint8_t dst_ip[NET6_LEN],
+	uint16_t src_port,
+	uint16_t dst_port,
+	uint8_t proto,
+	uint16_t flags
+) {
+	packet->mbuf =
+		make_mbuf6(src_ip, dst_ip, src_port, dst_port, proto, flags);
+	assert(packet->mbuf != NULL);
+	return parse_packet(packet);
+}
+
+int
+fill_packet(
+	struct packet *packet,
+	const uint8_t *src_ip,
+	const uint8_t *dst_ip,
+	uint16_t src_port,
+	uint16_t dst_port,
+	uint8_t transport_proto,
+	uint8_t network_proto,
+	uint16_t flags
+) {
+	if (network_proto == IPPROTO_IP) {
+		return fill_packet_net4(
+			packet,
+			src_ip,
+			dst_ip,
+			src_port,
+			dst_port,
+			transport_proto,
+			flags
+		);
+	} else if (network_proto == IPPROTO_IPV6) {
+		return fill_packet_net6(
+			packet,
+			src_ip,
+			dst_ip,
+			src_port,
+			dst_port,
+			transport_proto,
+			flags
+		);
+	} else {
+		return -1;
+	}
+}
+
+void
+free_packet(struct packet *packet) {
+	free(packet->mbuf);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+static void
+init_mbuf(struct rte_mbuf *m, struct packet_data *data, uint16_t buf_len) {
+	m->priv_size = 0;
+	m->buf_len = buf_len;
+	uint32_t mbuf_size = sizeof(struct rte_mbuf) + m->priv_size;
+
+	/* start of buffer is after mbuf structure and priv data */
+	m->buf_addr = (char *)m + mbuf_size;
+
+	// what for?
```

Comment:
I think it was just a precaution, but from the current point of view, it's not necessary